SAL annotation updates & fix warnings

dantraMSFT opened the pull request:
Various SAL annotations were either incorrect or not backed up by code. This PR address these issues.

NOTE: By default, Start-BuildNativeWindowsBinaries does not enable code analysis and issues detected by SAL annotations are not reported.  To identify these issues, run Start-BuildNativeWindowsBinaries to generate the solution and vcxproj files.  Launch a Visual Studio developer prompt,  cd to src\powershell-native and run msbuild manually. 

```
msbuild PowerShellNative.sln /p:RunCodeAnalysis=true /p:Configuration=RelWithDebInfo /p:Platform=x64
```

The following changes address all code analysis warnings:
* GetRegistryInfo in NativeMsh had incorrect out annotations, remove __opt
* Fix handling of various out parameters - check for non-null and initialize
* Update and Align  SAL annotations for GetFormattedErrorMessage overloads
* Allow PluginException to accept NULL message.

dantraMSFT committed:
SAL annotation updates and fix warnings
GetRegistryInfo in NativeMsh had incorrect out annotations, remove __opt
Fix handling of various out parameters - check for non-null and initialize
Udpate and Align  SAL annotations for GetFormattedErrorMessage overloads
Allow PluginException to accept NULL message.

daxian-dbw commented:
Note that the `psrp.windows` nuget package needs to be regenerated. Make sure that the binaries are signed before using them to create the nuget package.

dantraMSFT commented:
@TravisEz13: The failures in these two runs are unrelated to the change; they are timing issues.  I've created https://github.com/PowerShell/PowerShell/issues/5627 to track the issue.

dantraMSFT commented:
@daxian-dbw: I don't expect to be able to get the nuget package signed and published before I leave for vacation (EOD today).

daxian-dbw commented:
@SteveL-MSFT @TravisEz13 Can you please reivew this PR to see if it's needed for RC2 or GA?

daxian-dbw commented on src/powershell-native/nativemsh/pwrshcommon/pwrshcommon.cpp:
```diff
@@ -1238,31 +1238,47 @@ namespace NativeMsh
+
+        if (NULL != pwszMonadVersion)
+        {
+            pwszMonadVersion = NULL;
```
The two check above are assigning `NULL` to `*<pointer>` while this one is not. Is it intentional?

dantraMSFT replied:
Fixed

mirichmo commented on src/powershell-native/nativemsh/pwrshplugin/entrypoints.cpp:
```diff
@@ -299,6 +301,7 @@ unsigned int GetCLRVersionForPSVersion(int iPSMajorVersion,
 
 DWORD ReportOperationComplete(WSMAN_PLUGIN_REQUEST *requestDetails, DWORD errorCode)
 {
+    WCHAR szMessage[256] = L"\0";
```
Why did you make this code change? It looks unnecessary and potentially error-prone.

It is incorrect to provide a fixed-length buffer for the error string. The string is read from pwrshplugin.dll and may be of any length. GetFormattedErrorMessage correctly handles the arbitrary length, but this change will potentially truncate it.

dantraMSFT replied:
There is an assignment missing, which I'll add in the next commit; szMessage is used when GetFormmatedMessage fails.

mirichmo replied:
The assignment does not address my comment. This change is fundamentally flawed because it will potentially truncate the string. I do not see a benefit over the original code.

mirichmo commented on src/powershell-native/nativemsh/pwrshplugin/entrypoints.cpp:
```diff
@@ -307,11 +310,14 @@ DWORD ReportOperationComplete(WSMAN_PLUGIN_REQUEST *requestDetails, DWORD errorC
     result = WSManPluginOperationComplete(requestDetails, 0, errorCode, pwszErrorMessage);
 
-    if (NULL != pwszErrorMessage)
+    if (NULL != pwszErrorMessage && pwszErrorMessage != szMessage)
```
This change will conditionally leak memory. If pwszErrorMessage has a value, it was allocated by GetFormattedErrorMessage and must be freed (via delete[]).

dantraMSFT replied:
Fixed

mirichmo replied:
This has not been resolved yet.

mirichmo commented on src/powershell-native/nativemsh/pwrshplugin/pwrshclrhost.cpp:
```diff
@@ -270,7 +270,10 @@ unsigned int PowerShellClrWorker::LoadWorkerCallbackPtrs(
         PWSTR msg = NULL;
         exitCode = g_MANAGED_METHOD_RESOLUTION_FAILED;
-        GetFormattedErrorMessage(&msg, exitCode);
+        /* NOTE: dont' use a string literal for a fallback; PlugInException expects msg to allocated
```
This comment seems unnecessary. The behavior is documented in the class definition.

dantraMSFT replied:
I added this comment because the constructor annotations were wrong and, since the return value of GetFormmatedErrorMessage was being ignored, I added the comment to document the behavior as intentional.

mirichmo commented on src/powershell-native/nativemsh/pwrshplugin/pwrshplugin.h:
```diff
@@ -490,6 +490,11 @@ class PwrshPlugInMediator
         {
             iMajorVersion = iVPSMajorVersion;
 
+            if (NULL != wszMgdPlugInFileName)
```
This already occurs on line 671. It seems unnecessary to do it twice. Did the scanner require it?

dantraMSFT replied:
This check is added because wszMgdPlugInFileName was not being set in all paths; specifically error paths.

mirichmo commented on src/powershell-native/nativemsh/pwrshplugin/pwrshplugin.h:
```diff
@@ -768,7 +780,9 @@ class PwrshPlugInMediator
-                GetFormattedErrorMessage(&msg, g_OPTION_SET_NOT_COMPLY, g_BUILD_VERSION);
+                (void) GetFormattedErrorMessage(&msg, g_OPTION_SET_NOT_COMPLY, g_BUILD_VERSION);
+                /* NOTE: Passing possible NULL msg. PlugInExpecption requires allocated
+                   or NULL. Literal strings are not supported. */
```
Comment seems unnecessary here too

dantraMSFT committed:
Address PR feedback.

dantraMSFT commented:
@mirichmo: If you take a look at the code again, you'll see that szMessage is to populate a placeholder error when when GetFormattedErrorMessage fails. It does not truncate an arbitrary error string.

As far as a possible leak, there isn't one.

When GetFormattedErrorMessage returns zero; pwszErrorMessage is null and, szMessage is used instead; thus, the delete call only frees memory when pwszErrorMessage is non-null and not szMessage.

Are you seeing something that I have missed?

SteveL-MSFT commented on src/powershell-native/nativemsh/pwrshplugin/entrypoints.cpp:
```diff
@@ -299,6 +301,7 @@ unsigned int GetCLRVersionForPSVersion(int iPSMajorVersion,
 
 DWORD ReportOperationComplete(WSMAN_PLUGIN_REQUEST *requestDetails, DWORD errorCode)
 {
+    WCHAR szMessage[256] = L"\0";
```
After talking to Dan, `szMessage` contains a default error message in the case where the actual error message can't be allocated.  So `szMessage` should be called something else like `szFallbackErrorMessage` making this more clear.  Second, the 256 bytes is arbitrary.  Since we know the max length based on line 315 below, we should only allocate that much.

dantraMSFT replied:
Fixed.
Note that I left the stack buffer at 64.

SteveL-MSFT commented on src/powershell-native/nativemsh/pwrshplugin/entrypoints.cpp:
```diff
@@ -307,11 +310,15 @@ DWORD ReportOperationComplete(WSMAN_PLUGIN_REQUEST *requestDetails, DWORD errorC
+    if (GetFormattedErrorMessage(&pwszErrorMessage, errorCode) == 0)
+    {
+        swprintf_s(szMessage,_countof(szMessage), L"ReportOperationComplete: %d", errorCode);
+        pwszErrorMessage = szMessage;
```
Perhaps for readability of the code, it would be better to change this to call WSManPluginOperationComplete with szMessage and else use pwszErrorMessage?

This would also remove the change on line 321 below.

dantraMSFT replied:
Fixed.

dantraMSFT committed:
Rewrite ReportOperationComplete for clarity
Update GetFormattedErrorMessage to ensure all error paths return dwLength = 0

dantraMSFT commented:
@mirichmo, @SteveL-MSFT  
I've updated ReportOperationComplete to clarify the stack buffer usage.
After another review, I found that GetFormattedErrorMessage was still not reliably returning zero for all error paths.  I've rewritten it to be explicit in each case and avoid the nested If statements by leveraging the existing do/while(false)  construct and moving the localfree call for wszSystemErrorMessage to the end of the function.

dantraMSFT commented:
@TravisEz13 See https://github.com/PowerShell/PowerShell/issues/5802
